File: tictactoe.rb
def init_board
  board = []
  cell = '1'.ord
  3.times do
    row = []
    3.times do
      row << cell.chr
      cell += 1
    end
    board << row
  end
  board
end

def print_board(board)
  puts
  3.times do |i|
    puts " #{board[i][0]} | #{board[i][1]} | #{board[i][2]} "
    puts "---+---+---" if i != 2
  end
  puts
end

def check_win(board)
  3.times do |i|
    return true if board[i][0] == board[i][1] && board[i][1] == board[i][2]
    return true if board[0][i] == board[1][i] && board[1][i] == board[2][i]
  end
  return true if board[0][0] == board[1][1] && board[1][1] == board[2][2]
  return true if board[0][2] == board[1][1] && board[1][1] == board[2][0]
  false
end

def is_draw(board)
  3.times do |i|
    3.times do |j|
      return false if ('1'..'9').include?(board[i][j])
    end
  end
  true
end

def make_move(board, player, choice)
  target = choice.to_s
  3.times do |i|
    3.times do |j|
      if board[i][j] == target
        board[i][j] = player
        return true
      end
    end
  end
  false
end

def main
  player = 'X'
  game_over = false
  board = init_board
  puts "Tic Tac Toe (Player X vs Player O)"
  print_board(board)

  until game_over
    print "Player #{player}, enter a cell number (1-9): "
    input = gets.chomp
    begin
      choice = Integer(input)
    rescue ArgumentError
      puts "Invalid input. Exiting."
      break
    end

    if choice < 1 || choice > 9
      puts "Invalid input. Exiting."
      break
    end

    unless make_move(board, player, choice)
      puts "Cell already taken. Try again."
      next
    end

    print_board(board)

    if check_win(board)
      puts "Player #{player} wins!"
      game_over = true
    elsif is_draw(board)
      puts "It's a draw."
      game_over = true
    else
      player = (player == 'X' ? 'O' : 'X')
    end
  end
end

if __FILE__ == $0
  main
end

File: tictactoe.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func initBoard() [][]string {
	board := make([][]string, 3)
	cell := '1'
	for i := 0; i < 3; i++ {
		row := make([]string, 3)
		for j := 0; j < 3; j++ {
			row[j] = string(cell)
			cell++
		}
		board[i] = row
	}
	return board
}

func printBoard(board [][]string) {
	fmt.Println()
	for i := 0; i < 3; i++ {
		fmt.Printf(" %s | %s | %s \n", board[i][0], board[i][1], board[i][2])
		if i != 2 {
			fmt.Println("---+---+---")
		}
	}
	fmt.Println()
}

func checkWin(board [][]string) bool {
	// rows and columns
	for i := 0; i < 3; i++ {
		if board[i][0] == board[i][1] && board[i][1] == board[i][2] {
			return true
		}
		if board[0][i] == board[1][i] && board[1][i] == board[2][i] {
			return true
		}
	}
	// diagonals
	if board[0][0] == board[1][1] && board[1][1] == board[2][2] {
		return true
	}
	if board[0][2] == board[1][1] && board[1][1] == board[2][0] {
		return true
	}
	return false
}

func isDraw(board [][]string) bool {
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if board[i][j] >= "1" && board[i][j] <= "9" {
				return false
			}
		}
	}
	return true
}

func makeMove(board [][]string, player string, choice int) bool {
	target := strconv.Itoa(choice)
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if board[i][j] == target {
				board[i][j] = player
				return true
			}
		}
	}
	return false
}

func main() {
	player := "X"
	gameOver := false
	board := initBoard()
	fmt.Println("Tic Tac Toe (Player X vs Player O)")
	printBoard(board)

	scanner := bufio.NewScanner(os.Stdin)

	for !gameOver {
		fmt.Printf("Player %s, enter a cell number (1-9): ", player)
		if !scanner.Scan() {
			fmt.Println("Invalid input. Exiting.")
			break
		}
		input := strings.TrimSpace(scanner.Text())
		choice, err := strconv.Atoi(input)
		if err != nil {
			fmt.Println("Invalid input. Exiting.")
			break
		}

		if choice < 1 || choice > 9 {
			fmt.Println("Invalid input. Exiting.")
			break
		}

		if !makeMove(board, player, choice) {
			fmt.Println("Cell already taken. Try again.")
			continue
		}

		printBoard(board)

		if checkWin(board) {
			fmt.Printf("Player %s wins!\n", player)
			gameOver = true
		} else if isDraw(board) {
			fmt.Println("It's a draw.")
			gameOver = true
		} else {
			if player == "X" {
				player = "O"
			} else {
				player = "X"
			}
		}
	}
}
